src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        FreeTextEntries(dest);
+    }
+
+    return error;
+}
+
+} // namespace
+
+namespace chip {
+namespace Mdns {
+
+ServicePool::Entry::Entry() : mHasValue(false), mHasPending(false), mNodeId(0), mFabricId(0)
+{
+    mService.mTextEntryies  = nullptr;
+    mService.mTextEntrySize = 0;
+}
+
+ServicePool::Entry::Entry(Entry && rhs)
+{
+    MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::operator=(Entry && rhs)
+{
+    return MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::MoveFrom(Entry && rhs)
+{
+    mService    = rhs.mService;
+    mNodeId     = rhs.mNodeId;
+    mFabricId   = rhs.mFabricId;
+    mHasValue   = rhs.mHasValue;
+    mHasPending = rhs.mHasPending;
+
+    rhs.mHasValue   = false;
+    rhs.mHasPending = false;
+
+    return *this;
+}
+
+CHIP_ERROR ServicePool::Entry::Emplace(const MdnsService & service, uint64_t nodeId, uint64_t fabricId)
+{
+    CHIP_ERROR error;
+
+    Clear();
+    mService                          = service;
+    const_cast<uint64_t &>(mNodeId)   = nodeId;
+    const_cast<uint64_t &>(mFabricId) = fabricId;
+    mService.mTextEntryies            = nullptr;
+    error                             = CopyTextEntries(&mService, service);
+
+    if (error == CHIP_NO_ERROR)
+    {
+        mHasValue   = true;
+        mHasPending = false;
+    }
+    return error;
+}
+
+CHIP_ERROR ServicePool::Entry::Emplace(MdnsService && service, uint64_t nodeId, uint64_t fabricId)
+{
+    Clear();
+    mService               = service;
+    mNodeId                = nodeId;
+    mFabricId              = fabricId;
+    service.mTextEntryies  = nullptr;
+    service.mTextEntrySize = 0;
+    mHasValue              = true;
+    mHasPending            = false;
+
+    return CHIP_NO_ERROR;
+}
+
+void ServicePool::Entry::Clear()
+{
+    if (mHasValue)
+    {
+        FreeTextEntries(&mService);
+    }
+    mHasValue   = false;
+    mHasPending = false;
+}
+
+ServicePool::Entry::~Entry()
+{
+    Clear();
+}
+
+ServicePool::Entry & ServicePool::FindAvailableSlot(size_t hashValue)
+{
+    size_t i   = hashValue;
+    bool found = false;
+
+    do
+    {
+        if (!mEntries[i].mHasValue)
+        {
+            found = true;
+            break;
+        }
+
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    if (!found)
+    {
+        i = hashValue;
+    }
+
+    return mEntries[i];
+}
+
+CHIP_ERROR ServicePool::AddService(uint64_t nodeId, uint64_t fabricId, const MdnsService & service)
+{
+    return FindAvailableSlot(nodeId % kServicePoolCapacity).Emplace(service, nodeId, fabricId);
+}
+
+CHIP_ERROR ServicePool::AddService(uint64_t nodeId, uint64_t fabricId, MdnsService && service)
+{
+
+    return FindAvailableSlot(nodeId % kServicePoolCapacity).Emplace(std::move(service), nodeId, fabricId);
```

Comment:
`FindAvailableSlot` could take a node id, so the hashvalue computation would just live inside there, right?